File: client-sdk/java/sdk/src/main/java/com/vertexcache/sdk/protocol/command/SetCommand.java
package com.vertexcache.sdk.protocol.command;

import com.vertexcache.sdk.protocol.BaseCommand;
import com.vertexcache.sdk.protocol.CommandType;
import com.vertexcache.sdk.result.VertexCacheSdkException;

public class SetCommand extends BaseCommand<SetCommand> {

    private final String primaryKey;
    private final String value;
    private final String secondaryKey;
    private final String tertiaryKey;

    public SetCommand(String primaryKey, String value) throws VertexCacheSdkException {
        this(primaryKey, value, null, null);
    }

    public SetCommand(String primaryKey, String value, String secondaryKey) throws VertexCacheSdkException {
        this(primaryKey, value, secondaryKey, null);
    }

    public SetCommand(String primaryKey, String value, String secondaryKey, String tertiaryKey) throws VertexCacheSdkException {

        if(primaryKey == null || primaryKey.isBlank()) {
            throw new VertexCacheSdkException("Missing Primary Key");
        }

        if(value == null || value.isBlank()) {
            throw new VertexCacheSdkException("Missing Value");
        }

        this.primaryKey = primaryKey;
        this.value = value;
        this.secondaryKey = secondaryKey;
        this.tertiaryKey = tertiaryKey;
    }

    @Override
    protected String buildCommand() {
        StringBuilder sb = new StringBuilder();
        sb.append(CommandType.SET).append(BaseCommand.COMMAND_SPACER)
                .append(primaryKey).append(BaseCommand.COMMAND_SPACER)
                .append(value);

        if (secondaryKey != null && !secondaryKey.isBlank()) {
            sb.append(" ").append(CommandType.IDX1).append(" ").append(secondaryKey);
        }

        if (tertiaryKey != null && !tertiaryKey.isBlank()) {
            sb.append(" ").append(CommandType.IDX2).append(" ").append(tertiaryKey);
        }

        return sb.toString();
    }

    protected void parseResponse(String responseBody) {
        if(!responseBody.equalsIgnoreCase("OK")) {
            this.setFailure("OK Not received");
        } else {
            this.setSuccess();
        }
    }
}

File: client-sdks/rust/sdk/command/impls/set_command.rs
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// ------------------------------------------------------------------------------

use crate::comm::client_connector::ClientConnector;
use crate::command::command::Command;
use crate::command::command_type::CommandType;
use crate::model::vertex_cache_sdk_exception::VertexCacheSdkException;

pub struct SetCommand {
    pub primary_key: String,
    pub value: String,
    pub secondary_key: Option<String>,
    pub tertiary_key: Option<String>,
    pub success: bool,
    pub status_message: String,
}

impl SetCommand {
    pub fn new(
        primary_key: String,
        value: String,
        secondary_key: Option<String>,
        tertiary_key: Option<String>,
    ) -> Result<Self, VertexCacheSdkException> {
        if primary_key.trim().is_empty() {
            return Err(VertexCacheSdkException::new("Missing Primary Key"));
        }

        if value.trim().is_empty() {
            return Err(VertexCacheSdkException::new("Missing Value"));
        }

        if let Some(sk) = &secondary_key {
            if sk.trim().is_empty() {
                return Err(VertexCacheSdkException::new("Secondary key can't be empty when used"));
            }
        }

        if let (Some(sk), Some(tk)) = (&secondary_key, &tertiary_key) {
            if !sk.trim().is_empty() && tk.trim().is_empty() {
                return Err(VertexCacheSdkException::new("Tertiary key can't be empty when used"));
            }
        }

        Ok(Self {
            primary_key,
            value,
            secondary_key,
            tertiary_key,
            success: false,
            status_message: String::new(),
        })
    }

    fn build_command(&self) -> String {
        let mut cmd = format!("SET {} {}", self.primary_key, self.value);

        if let Some(sk) = &self.secondary_key {
            if !sk.trim().is_empty() {
                cmd.push_str(&format!(" IDX1 {}", sk));
            }
        }

        if let Some(tk) = &self.tertiary_key {
            if !tk.trim().is_empty() {
                cmd.push_str(&format!(" IDX2 {}", tk));
            }
        }

        cmd
    }
}

impl Command for SetCommand {
    fn execute(&mut self, connector: &mut ClientConnector) {
        let cmd = self.build_command();

        match connector.send(&cmd) {
            Ok(response) => {
                let cleaned = response.trim().trim_start_matches('+');
                self.status_message = cleaned.to_string();
                self.success = cleaned.eq_ignore_ascii_case("OK");
            }
            Err(e) => {
                self.success = false;
                self.status_message = format!("SET command failed: {}", e);
            }
        }
    }

    fn command_type(&self) -> CommandType {
        CommandType::Set
    }

    fn is_success(&self) -> bool {
        self.success
    }

    fn status_message(&self) -> &str {
        &self.status_message
    }
}
